### PR TITLE
REF: stronger typing in libconversion

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/_libs/algos_common_helper.pxi.in [hunk 1]
```diff
@@ -65,7 +65,8 @@ def ensure_{{name}}(object arr, copy=True):
         if (<ndarray>arr).descr.type_num == NPY_{{c_type}}:
             return arr
         else:
-            return arr.astype(np.{{dtype}}, copy=copy)
+            # equiv: arr.astype(np.{{dtype}}, copy=copy)
+            return cnp.PyArray_Cast(<ndarray>arr, cnp.NPY_{{c_type}})
```
Comment (mroeschke) [inline pandas/_libs/algos_common_helper.pxi.in:69]: If I'm interpreting this correctly now it always copies? https://numpy.org/devdocs/reference/c-api/array.html#c.PyArray_Cast

If that's correct can `copy=True` be removed in the signature?

Comment (jbrockmendel) [reply pandas/_libs/algos_common_helper.pxi.in:69]: We only get here when the dtypes don't match, so i think a copy needs to be made regardless of the copy keyword.

Comment (mroeschke) [reply pandas/_libs/algos_common_helper.pxi.in:69]: Got it, so can the `copy=True` in the signature be removed then as well?

Comment (jbrockmendel) [reply pandas/_libs/algos_common_helper.pxi.in:69]: i want to investigate whether the fact that we _dont_ make a copy in the matching-dtype case a few lines up causes a subtle bug anywhere.  If not, then sure.

Comment (mroeschke) [reply pandas/_libs/algos_common_helper.pxi.in:69]: Were you thinking of investigating in this PR or a follow up?

Comment (jbrockmendel) [reply pandas/_libs/algos_common_helper.pxi.in:69]: was planning to do that separately

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**